exercises/nondet_monad_cont/descr.md

```diff
@@ -0,0 +1,521 @@
+# Implementing Nondeterminism with Continuations
+
+In this exercise, we build an implementation of the nondeterminism
+monad. This monad admits several possible implementations; the one
+that we choose here is a continuation-based implementation where a
+computation receives two continuations, a *success continuation*
+and a *failure continuation*, which it invokes to indicate the
+presence or absence of more results.
+
+## The Nondeterminism Monad
+
+When searching for the solution of a problem, one must typically
+explore multiple choices. If a series of choices lead to a failure
+(a dead end), then one must backtrack and explore another avenue.
+
+There are a number of ways in which nondeterminism and backtracking
+can be implemented. Regardless of which implementation mechanism is
+chosen, it is desirable to hide it behind an abstraction barrier and
+present the end user with **a simple API for constructing an executing
+nondeterministic computations**.
+
+This API is known as the **nondeterminism monad**. It offers the
+following key elements:
+
+* A type `'a m`, the type of computations that yield results of type
+ `'a`.
+
+* A number of constructor functions for constructing computations,
+  such as `fail: 'a m`, which represents failure, and `choose: 'a m
+  -> 'a m -> 'a m`, which expresses a nondeterministic choice
+  between two computations.
+
+* A single observation function, `sols: 'a m -> 'a Seq.t`, which
+  converts a computation to a sequence of results, thereby allowing
+  the user to execute this computation and observe its results. The
+  name `sols` stands for `solutions`.
+
+A monad can be thought of as a **mini-programming language** where
+computations are first-class citizens: we have a type of
+computations, ways of building computations, and a way of executing
+computations.
+
+A computation in the nondeterminism monad can produce zero, one, or
+more results. Indeed, a computation that fails produces zero
+results. A computation that succeeds normally produces one result.
+A computation that uses `choose` can produce more than one result.
+It is in fact possible to construct computations that produce an
+infinite number of results!
+
+Thus, a useful way to think of a computation is as **a sequence of
```

Comment:
This sentence is weird :-)

Reply:
I was referring to "as a sequence of results, that is, a sequence of results." :-)

Reply:
Oh, I see. Thanks; fixed.